### PR TITLE
Wb 1056 edge subdivision ui

### PR DESCRIPTION
**Pull request by MartinImgSpc**
New widget allowing a min max range + editable UI range. As per Gabriel request there is no automatic expansion of the range setting section.

### CODE REVIEW

File: Editor/EditorCore/EditorGUIUtility.cs [hunk 1]
```diff
@@ -311,6 +311,81 @@ public static float FreeSlider(GUIContent content, float value, float min, float
...
+            if (content != null)
+                GUI.Label(labelRect, content);
+
+            EditorGUI.BeginChangeCheck();
+
+            int controlID = GUIUtility.GetControlID(FocusType.Passive, sliderRect);
+            float tmp = value;
+            float tmpUIMin = uiMin;
+            float tmpUIMax = uiMax;
+            tmp = GUI.Slider(sliderRect, tmp, 0f, tmpUIMin, tmpUIMax, GUI.skin.horizontalSlider, (!EditorGUI.showMixedValue) ? GUI.skin.horizontalSliderThumb : "SliderMixed", true, controlID);
```
Comment (vkovec) [inline Editor/EditorCore/EditorGUIUtility.cs]: is this the same as https://docs.unity3d.com/ScriptReference/GUI.HorizontalSlider.html?

File: Editor/EditorCore/EditorGUIUtility.cs [hunk 2]
```diff
@@ -311,6 +311,81 @@ public static float FreeSlider(GUIContent content, float value, float min, float
...
+
+        /**
+         * Similar to EditorGUILayoutUtility.Slider, except this allows for values outside of the uiMin/uiMax bounds via the int field.
+         * Contrary to the FreeSlider however it has a hard range defined with min and max.
+         */
+        public static int FreeSliderWithRange(GUIContent content, int value, int min, int max, ref int uiMin, ref int uiMax, ref bool expanded)
+        {
+            float pixelsPerPoint = 1f;
+
+#if !UNITY_4_7 && !UNITY_5_0 && !UNITY_5_3
```
[the remark this answers is not in the document]
Comment (karl-) [reply Editor/EditorCore/EditorGUIUtility.cs]: nope, these can be removed. we only support back to 2018.3.

**Comment by MartinImgSpc** [comment]
> Issues:
> 
> 1. If user types a value above the current user-controllable max, reset the user-controllable max to that value, don't hide the slider thumb
> 2. If user modifiers the user-controllable max, and it is lower than the current action value, drop the action value to that user-typed max, don't hide the slider thumb
> 3. If user types a value below the current user-controllable min, reset the user controllable min to that value, don't hide the slider thumb
> 4. If user modifiers the user-controllable min, and it is higher than the current action value, increase the action value to that user-typed min, do not hide the slider thumb
> 5. Missing padding between the label, fold out, and slider - should be at least 6px between each

Will look at those after SIGGRAPH thanks for your feedback!

**Comment by gabrielw-us** [comment]
Some items, noted it slack:
- typing into one value field does not update other fields until hitting "enter", this feels odd compared to other areas of Unity, ie on a Material, set the "Smoothness" via value field and the slider updates as you type, vs here the slider and any dependent fields do not update until hitting enter. Is it possible to catch this and keep the other fields + slider updated?
- foldout arrow seems to have incorrect click zone, I need to click a few pixels behind it, clicking directly on it does not work for me
- need same visual pixel padding between min and max fields as between other elements ie the slider horizontal bar and it's field

**Comment by MartinImgSpc** [comment]
> Some items, noted it slack:
> 
> * typing into one value field does not update other fields until hitting "enter", this feels odd compared to other areas of Unity, ie on a Material, set the "Smoothness" via value field and the slider updates as you type, vs here the slider and any dependent fields do not update until hitting enter. Is it possible to catch this and keep the other fields + slider updated?
> * foldout arrow seems to have incorrect click zone, I need to click a few pixels behind it, clicking directly on it does not work for me
> * need same visual pixel padding between min and max fields as between other elements ie the slider horizontal bar and it's field

With regards to the enter unfortunately this is the only way I can know when to do the validation and compensation. Otherwise if I had a Minimum UI value of 4. And started typing 12 for value.I would end up with 1 in Minimum UI value.

I will look at the padding and click zone.

**Comment by gabrielw-us** [comment]
Sorry, one more tiny issue, I should have noticed - there should be 2px height padding between the min/max and slider value. Currently they overlap.

**Pull request merged**